File: src/main/java/io/github/saneea/dvh/FeatureContext.java
package io.github.saneea.dvh;

import java.util.stream.Collectors;
import java.util.stream.Stream;

public class FeatureContext {

	private final FeatureContext parent;
	private final String featureName;
	private final String[] args;

	public FeatureContext(FeatureContext parent, String featureName, String[] args) {
		this.parent = parent;
		this.featureName = featureName;
		this.args = args;
	}

	public FeatureContext getParent() {
		return parent;
	}

	public String getFeatureName() {
		return featureName;
	}

	public String[] getArgs() {
		return args;
	}

	public String getFeaturesChainString() {
		return getFeaturesChain().collect(Collectors.joining(" "));
	}

	public Stream<String> getFeaturesChain() {
		return Stream.concat(//
				parent != null//
						? parent.getFeaturesChain()//
						: Stream.empty(), //
				Stream.of(getFeatureName()));
	}

}

File: src/main/kotlin/io/github/saneea/dvh/FeatureContext.kt
package io.github.saneea.dvh

import java.util.stream.Collectors
import java.util.stream.Stream

class FeatureContext(
    private val parent: FeatureContext?,
    val featureName: String,
    val args: Array<String>
) {
    val featuresChainString: String
        get() = featuresChain.collect(Collectors.joining(" "))

    val featuresChain: Stream<String>
        get() = Stream.concat(
            parent?.featuresChain ?: Stream.empty(),
            Stream.of(featureName)
        )
}